cloudfoundry/resource_cf_service_instance.go

```diff
@@ -157,26 +168,70 @@ func resourceServiceInstanceUpdate(d *schema.ResourceData, meta interface{}) (er
 		tags = append(tags, v.(string))
 	}
 
-	_, err = sm.UpdateServiceInstance(id, name, servicePlan, params, tags)
-	return err
+	if _, err = sm.UpdateServiceInstance(id, name, servicePlan, params, tags); err != nil {
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// Check whetever service_instance exists and is in state 'succeeded'
+	if err = sm.WaitServiceInstanceTo("update", id); err != nil {
+		return err
+	}
+
+	return nil
 }
 
 func resourceServiceInstanceDelete(d *schema.ResourceData, meta interface{}) (err error) {
 
 	session := meta.(*cfapi.Session)
+	id := d.Id()
+
 	if session == nil {
 		return fmt.Errorf("client is nil")
 	}
 	session.Log.DebugMessage("begin resourceServiceInstanceDelete")
 
 	sm := session.ServiceManager()
 
-	err = sm.DeleteServiceInstance(d.Id())
-	if err != nil {
+	if err = sm.DeleteServiceInstance(id); err != nil {
+		return err
+	}
+
+	// Check whether service_instance has been indeed deleted (sometimes takes very long))
+	if err = sm.WaitDeletionServiceInstance(id); err != nil {
 		return err
 	}
 
 	session.Log.DebugMessage("Deleted Service Instance : %s", d.Id())
 
 	return nil
 }
+
+func resourceServiceInstanceImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	session := meta.(*cfapi.Session)
+
+	if session == nil {
+		return nil, fmt.Errorf("client is nil")
+	}
+
+	sm := session.ServiceManager()
+
+	serviceinstance, err := sm.ReadServiceInstance(d.Id())
+
+	if err != nil {
+		return nil, err
+	}
+
+	d.Set("name", serviceinstance.Name)
```

Comment:
Most of the importers contributed by @ArthurHlt in https://github.com/mevansam/terraform-provider-cf/pull/14/files use the `ImportStatePassthrough` pattern. Would it also be applicable here ?

Reply:
Thanks @gberche-orange . I have added that pattern, and fix the broken import, and extended the test for that.